redis/server.cc

```diff
@@ -0,0 +1,259 @@
+    }
+    auto nr_conn = make_lw_shared<size_t>(0);
+    auto nr_conn_total = _connections_list.size();
+    logging.debug("redis_server: shutdown connection nr_total={}", nr_conn_total);
+    return parallel_for_each(_connections_list.begin(), _connections_list.end(), [nr_conn, nr_conn_total] (auto&& c) {
+        return c.shutdown().then([nr_conn, nr_conn_total] {
+            logging.debug("redis_server: shutdown connection {} out of {} done", ++(*nr_conn), nr_conn_total);
+        });
+    }).then([this] {
+        return std::move(_stopped);
+    });
+}
+
+future<> redis_server::listen(socket_address addr, std::shared_ptr<seastar::tls::credentials_builder> creds, bool keepalive) {
+    listen_options lo;
+    lo.reuse_address = true;
+    server_socket ss;
+    try {
+        ss = creds
+          ? seastar::tls::listen(creds->build_server_credentials(), addr, lo)
+          : engine().listen(addr, lo);
+    } catch (...) {
+        throw std::runtime_error(sprint("Redis server error while listening on %s -> %s", addr, std::current_exception()));
+    }
+    _listeners.emplace_back(std::move(ss));
+    _stopped = when_all(std::move(_stopped), do_accepts(_listeners.size() - 1, keepalive, addr)).discard_result();
+    return make_ready_future<>();
+}
+
+future<> redis_server::do_accepts(int which, bool keepalive, socket_address server_addr) {
+    return repeat([this, which, keepalive, server_addr] {
+        ++_connections_being_accepted;
+        return _listeners[which].accept().then_wrapped([this, which, keepalive, server_addr] (future<accept_result> f_cs_sa) mutable {
+            --_connections_being_accepted;
+            if (_stopping) {
+                f_cs_sa.ignore_ready_future();
+                maybe_idle();
+                return stop_iteration::yes;
+            }
+            auto cs_sa = f_cs_sa.get0();
+            auto fd = std::move(cs_sa.connection);
+            auto addr = std::move(cs_sa.remote_address);
+            fd.set_nodelay(true);
+            fd.set_keepalive(keepalive);
+            auto conn = make_shared<connection>(*this, server_addr, std::move(fd), std::move(addr));
+            ++_connects;
+            ++_connections;
+            (void)conn->process().then_wrapped([this, conn] (future<> f) {
+                --_connections;
+                try {
+                    f.get();
+                } catch (...) {
+                    logging.debug("connection error: {}", std::current_exception());
+                }
+            });
+            return stop_iteration::no;
+        }).handle_exception([] (auto ep) {
+            logging.debug("accept failed: {}", ep);
+            return stop_iteration::no;
+        });
+    });
+}
+
+future<redis_server::result> redis_server::connection::process_request_one(redis::request&& request, redis::redis_options& opts, service_permit permit) {
+    return futurize_apply([this, request = std::move(request), &opts, permit] () mutable {
```

Comment:
Why do you need futurize_apply and a lambda here? I'm not sure I understand what you're trying to do...